File: frontend/lib/model/nghiphep/nghi_phep.dart
class NghiPhep {
  final int? maNghiPhep;
  final int maNV;
  final String tuNgay; // yyyy-MM-dd
  final String denNgay; // yyyy-MM-dd
  final int? soNgay;
  final String loaiNghi; // PHEP_NAM, PHEP_OM, PHEP_THAI_SAN, PHEP_LE
  final String? lyDo;
  final String trangThai; // CHO_DUYET, DA_DUYET, TU_CHOI
  final int? nguoiDuyet;
  final String? ngayDuyet; // yyyy-MM-dd HH:mm:ss
  final String? ghiChuDuyet;
  final String? ngayGui; // yyyy-MM-dd HH:mm:ss
  final String? ngayCapNhat; // yyyy-MM-dd HH:mm:ss
  final bool daXoa;

  NghiPhep({
    this.maNghiPhep,
    required this.maNV,
    required this.tuNgay,
    required this.denNgay,
    this.soNgay,
    required this.loaiNghi,
    this.lyDo,
    this.trangThai = 'CHO_DUYET',
    this.nguoiDuyet,
    this.ngayDuyet,
    this.ghiChuDuyet,
    this.ngayGui,
    this.ngayCapNhat,
    this.daXoa = false,
  });

  factory NghiPhep.fromJson(Map<String, dynamic> json) {
    return NghiPhep(
      maNghiPhep: json['maNghiPhep'],
      maNV: json['maNV'],
      tuNgay: json['tuNgay'],
      denNgay: json['denNgay'],
      soNgay: json['soNgay'],
      loaiNghi: json['loaiNghi'],
      lyDo: json['lyDo'],
      trangThai: json['trangThai'] ?? 'CHO_DUYET',
      nguoiDuyet: json['nguoiDuyet'],
      ngayDuyet: json['ngayDuyet'],
      ghiChuDuyet: json['ghiChuDuyet'],
      ngayGui: json['ngayGui'],
      ngayCapNhat: json['ngayCapNhat'],
      daXoa: json['daXoa'] ?? false,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'maNghiPhep': maNghiPhep,
      'maNV': maNV,
      'tuNgay': tuNgay,
      'denNgay': denNgay,
      'soNgay': soNgay,
      'loaiNghi': loaiNghi,
      'lyDo': lyDo,
      'trangThai': trangThai,
      'nguoiDuyet': nguoiDuyet,
      'ngayDuyet': ngayDuyet,
      'ghiChuDuyet': ghiChuDuyet,
      'ngayGui': ngayGui,
      'ngayCapNhat': ngayCapNhat,
      'daXoa': daXoa,
    };
  }

  // Helper methods
  DateTime get tuNgayAsDateTime => DateTime.parse(tuNgay);
  DateTime get denNgayAsDateTime => DateTime.parse(denNgay);

  DateTime? get ngayDuyetAsDateTime {
    if (ngayDuyet == null) return null;
    return DateTime.parse(ngayDuyet!);
  }

  DateTime? get ngayGuiAsDateTime {
    if (ngayGui == null) return null;
    return DateTime.parse(ngayGui!);
  }

  String get loaiNghiDisplay {
    switch (loaiNghi) {
      case 'PHEP_NAM':
        return 'Phép năm';
      case 'PHEP_OM':
        return 'Phép ốm';
      case 'PHEP_THAI_SAN':
        return 'Phép thai sản';
      case 'PHEP_LE':
        return 'Phép lễ';
      default:
        return loaiNghi;
    }
  }

  String get trangThaiDisplay {
    switch (trangThai) {
      case 'CHO_DUYET':
        return 'Chờ duyệt';
      case 'DA_DUYET':
        return 'Đã duyệt';
      case 'TU_CHOI':
        return 'Từ chối';
      default:
        return trangThai;
    }
  }

  int get soNgayTinhToan {
    return denNgayAsDateTime.difference(tuNgayAsDateTime).inDays + 1;
  }

  bool get isDuocDuyet => trangThai == 'DA_DUYET';
  bool get isBiTuChoi => trangThai == 'TU_CHOI';
  bool get isChoDuyet => trangThai == 'CHO_DUYET';

  NghiPhep copyWith({
    int? maNghiPhep,
    int? maNV,
    String? tuNgay,
    String? denNgay,
    int? soNgay,
    String? loaiNghi,
    String? lyDo,
    String? trangThai,
    int? nguoiDuyet,
    String? ngayDuyet,
    String? ghiChuDuyet,
    String? ngayGui,
    String? ngayCapNhat,
    bool? daXoa,
  }) {
    return NghiPhep(
      maNghiPhep: maNghiPhep ?? this.maNghiPhep,
      maNV: maNV ?? this.maNV,
      tuNgay: tuNgay ?? this.tuNgay,
      denNgay: denNgay ?? this.denNgay,
      soNgay: soNgay ?? this.soNgay,
      loaiNghi: loaiNghi ?? this.loaiNghi,
      lyDo: lyDo ?? this.lyDo,
      trangThai: trangThai ?? this.trangThai,
      nguoiDuyet: nguoiDuyet ?? this.nguoiDuyet,
      ngayDuyet: ngayDuyet ?? this.ngayDuyet,
      ghiChuDuyet: ghiChuDuyet ?? this.ghiChuDuyet,
      ngayGui: ngayGui ?? this.ngayGui,
      ngayCapNhat: ngayCapNhat ?? this.ngayCapNhat,
      daXoa: daXoa ?? this.daXoa,
    );
  }
}

File: backend/src/main/java/com/example/worktrack/model/nghiphep/NghiPhep.java
package com.example.worktrack.model.nghiphep;

import com.fasterxml.jackson.annotation.JsonFormat;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDate;
import java.time.LocalDateTime;

@Entity
@Table(name = "nghi_phep")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class NghiPhep {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "ma_nghi_phep")
    private Integer maNghiPhep;

    // Mã nhân viên - khóa ngoại đến bảng nhan_vien
    @Column(name = "manv", nullable = false)
    private Integer maNV;

    @Column(name = "tu_ngay", nullable = false)
    @JsonFormat(pattern = "yyyy-MM-dd")
    private LocalDate tuNgay;

    @Column(name = "den_ngay", nullable = false)
    @JsonFormat(pattern = "yyyy-MM-dd")
    private LocalDate denNgay;

    @Column(name = "so_ngay", nullable = false)
    private Integer soNgay;

    @Column(name = "loai_nghi", length = 50, nullable = false)
    private String loaiNghi; // PHEP_NAM, PHEP_OM, PHEP_THAI_SAN, PHEP_LE

    @Column(name = "ly_do", columnDefinition = "TEXT")
    private String lyDo;

    @Column(name = "trang_thai", length = 20, nullable = false)
    private String trangThai = "CHO_DUYET"; // CHO_DUYET, DA_DUYET, TU_CHOI

    @Column(name = "nguoi_duyet")
    private Integer nguoiDuyet;

    @Column(name = "ngay_duyet")
    @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")
    private LocalDateTime ngayDuyet;

    @Column(name = "ghi_chu_duyet", columnDefinition = "TEXT")
    private String ghiChuDuyet;

    @Column(name = "ngay_gui", nullable = false)
    @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")
    private LocalDateTime ngayGui;

    @Column(name = "ngay_cap_nhat")
    @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")
    private LocalDateTime ngayCapNhat;

    @Column(name = "da_xoa", nullable = false)
    private Boolean daXoa = false;

    @PrePersist
    protected void onCreate() {
        ngayGui = LocalDateTime.now();
        ngayCapNhat = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        ngayCapNhat = LocalDateTime.now();
    }
}